### PR TITLE
Interfaces for thesaurus datatype

### PR DESCRIPTION
**Pull request by CascadingRadium**
- Add interfaces to abstract the thesaurus and its helper iterator methods.
- Extend Fuzzy and Regex FieldDict interfaces to return abstracted automatons for 
  calculating Damerau-Levenshtein distance and regex term matching, respectively,
  based on the original term/pattern using which these automatons were built.
- Add interfaces for special synonym documents and synonym fields. These interfaces allow 
  differentiation between synonym documents and regular documents during processing in the 
  index.

### CODE REVIEW

File: index.go [hunk 1]
```diff
@@ -105,12 +105,21 @@ type CopyReader interface {
 	CloseCopyReader() error
 }
 
+type RegexAutomaton interface {
+	MatchesRegex(string) bool
+}
```
Comment (abhinavdangeti) [inline index.go:110]: Add an empty line after.

File: index.go [hunk 2]
```diff
@@ -252,3 +261,27 @@ type IndexBuilder interface {
 	Index(doc Document) error
 	Close() error
 }
+
+type SynonymTermReader interface {
```
Comment (abhinavdangeti) [inline index.go:265]: Let's rename this to ThesaurusTermReader to align in naming?

File: index.go (hunk 1 above)
Comment (CascadingRadium) [reply index.go:110]: ack

File: index.go (hunk 2 above)
Comment (CascadingRadium) [reply index.go]: ack

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**